add support for TLS mongo connections

gchickma opened the pull request:
Adding support for clustered and TLS connections to MongoDB.

Intent is global.database.mongodb.host is used without global.database.mongodb.port, ie. the entire, host replicaSet is specified (including ports) in the string.

Also need a new TLS enable + certificate property for adding to the new certs secret containing a base64 encoded value

tlawrie commented:
Hi @gchickma can you create a matching GitHub issue and drop into 3.5.1 or 3.6 depending

gchickma commented:
Hey @tlawrie here's the GitHub issue https://github.com/boomerang-io/roadmap/issues/296

tlawrie commented on bmrg-flow/Chart.yaml:
```diff
@@ -2,7 +2,7 @@ apiVersion: v2
 type: application
-version: 4.11.3
+version: 4.11.4
 appVersion: 3.5.0
```
Will need to update this to 3.5.1 or 3.6.0 depending on when its going out officially.

gchickma replied:
Sure, I can update appVersion to 3.5.1

tlawrie commented on bmrg-flow/templates/configmap.yaml:
```diff
@@ -12,7 +12,11 @@ metadata:
 data:
   application.properties: |
     server.port=8080
+    {{- if $.Values.global.database.mongodb.tls.enable }}
+    spring.data.mongodb.uri=mongodb://{{ $.Values.global.database.mongodb.user }}:${MONGODB_PASSWORD}@{{ $.Values.global.database.mongodb.host }}
+    {{- else }}
     spring.data.mongodb.uri=mongodb://{{ $.Values.global.database.mongodb.user }}:${MONGODB_PASSWORD}@{{ $.Values.global.database.mongodb.host }}:{{ $.Values.global.database.mongodb.port }}/boomerang
+    {{- end }}
```
Could we not always send the port through?

gchickma replied:
Technically I need another flag that can be used to switch replicaSet enable/disable.  TLS is really just for whether we need to import a certificate into the cacerts keystore.  I will update this.

tlawrie replied:
Well we also dont want to add to many flags if we don't need them.

Can you just wrap the : and port with an if empty check don't add. So if you don't want to use, don't provide.

tlawrie replied:
Hey @gchickma have you thought about this one. It might be something like this (but I haven't valided)

```
spring.data.mongodb.uri=mongodb://{{ $.Values.global.database.mongodb.user }}:${MONGODB_PASSWORD}@{{ $.Values.global.database.mongodb.host }}{{ if $.Values.global.database.mongodb.port }}:{{ $.Values.global.database.mongodb.port }}{{ end }}/boomerang
```

tlawrie replied:
So ive done a big of search of other helm charts. I would say majority of them use the design of split parameters i.e. host / port / database / credentials / extraArgs vs url. I think URL makes it hard when dealing with credentials from secret though because in some parts of the yamls (like the deployment-service) it pulls the credentials from environment variables.

if we can, lets go with one long embedded if statements.

lets add `extraArgs` as a parameter.

tlawrie replied:
Updated schema

```
      host: bmrg-mdb001-ibm-mongodb-dev
      port: 27017
      database: boomerang
      args:
      user: boomerang
      password: passw0rd
      secretName:
```

tlawrie commented on bmrg-flow/templates/deployment-service.yaml:
```diff
@@ -97,6 +97,15 @@ spec:
             mountPath: /opt/newrelic
           {{- end }}
           {{- if and (eq $k "controller") (eq $values.workers.logging.type "elastic" ) }}
+          {{- if $.Values.database.mongodb.tls.enable }}
```
Can't we use a single statement with an or for the wrapper and then adjust what files are mounted?

gchickma replied:
Not really,  we need to know whether we need to import certificates for one or the other or both and it has to be done in single block given the java service is started after the keystore imports.

tlawrie replied:
Well in the case the IF statement should be combined and then wrap some of the command line with individual IF additions.

gchickma replied:
Actually yes, this one can be combined.

tlawrie replied:
Im not seeing the change in the above code

tlawrie commented on bmrg-flow/templates/deployment-service.yaml:
```diff
@@ -123,7 +143,13 @@ spec:
       {{- if and (eq $k "controller") (eq $values.workers.logging.type "elastic" ) }}
       - name: secret-volume
         secret:
-          secretName: {{ $values.workers.logging.secret }}
+          secretName: {{ include "bmrg.name" (dict "context" $ "tier" "certs" ) }}
+      {{- else }}
+      {{- if $.Values.database.mongodb.tls.enable }}
+      - name: secret-volume
+        secret:
+          secretName: {{ include "bmrg.name" (dict "context" $ "tier" "certs" ) }}
+      {{- end }}
```
Naming them both secret-volume is potentially going to lead to a collision if both MongoDB TLS and Elastic are enabled at the same time

tlawrie commented on bmrg-flow/templates/deployment-service.yaml:
```diff
@@ -123,7 +143,13 @@ spec:
       {{- if and (eq $k "controller") (eq $values.workers.logging.type "elastic" ) }}
       - name: secret-volume
         secret:
-          secretName: {{ $values.workers.logging.secret }}
+          secretName: {{ include "bmrg.name" (dict "context" $ "tier" "certs" ) }}
+      {{- else }}
+      {{- if $.Values.database.mongodb.tls.enable }}
```
The IF statement should be a single IF statement with an OR

gchickma replied:
As per previous comment, we need to know whether we need to mount secret for one or the other or both.

tlawrie replied:
Were you able to combine it? im not seeing the change come through

tlawrie commented:
Hey @gchickma few comments. 

Also do you have an example value of this
> Intent is global.database.mongodb.host is used without global.database.mongodb.port, ie. the entire, host replicaSet is specified (including ports) in the string.

The charts README would be a good spot to update as well as the installation notes in the docs

gchickma commented:
@tlawrie What's your thoughts on how to best handle replicaSets and other parameters?

tlawrie commented:
What do you mean, handle replicaSets and other parameters?

tlawrie commented on bmrg-flow/templates/secrets-cert.yaml:
```diff
@@ -0,0 +1,17 @@
+{{- $platform := "bmrg" -}}
+{{- $product := "flow" -}}
+{{- $tier := "certs" -}}
+apiVersion: v1
+kind: Secret
+metadata:
+  name: {{ include "bmrg.name" (dict "context" $ "tier" $tier ) }}
+  labels:
+    {{- include "bmrg.labels.chart" (dict "context" $ "tier" $tier ) | nindent 4 }}
+type: Opaque
+data:
+  {{- if (eq $values.workers.logging.type "elastic" ) }}
+  ca.crt: {{ $.Values.workers.logging.secret }}
+  {{- end }}
+  {{- if $.Values.global.database.mongodb.tls.enable }}
+  mongodb.crt: {{ $.Values.global.database.mongodb.tls.cert }}
+  {{- end }}
```
Hey @gchickma I am not sure I understand this file. I might be mistaken but I thought `$.Values.workers.logging.secret` was a reference to the secret file, not the contents to be put into a Secret file.

tlawrie commented on bmrg-flow/templates/deployment-service.yaml:
```diff
@@ -120,10 +120,10 @@ spec:
       {{- end }}
-      {{- if and (eq $k "controller") (eq $values.workers.logging.type "elastic" ) }}
+      {{- if or (and (eq $k "controller") (eq $values.workers.logging.type "elastic" )) ( $.Values.database.mongodb.tls.enable ) }}
       - name: secret-volume
         secret:
-          secretName: {{ $values.workers.logging.secret }}
+          secretName: {{ include "bmrg.name" (dict "context" $ "tier" "certs" ) }}
```
@gchickma this relates to the question on the secret file. We shouldn't be copying and storing the contents, it should just reference an existing secret created by the installation of Elastic or Mongo

gchickma commented:
@tlawrie Made further changes for review:
- Removed Elastic config from secrets-cert.yaml
- Updated configmap.yaml to handle optional extraParams parameter
- Added extraParams to values.yaml
- Mounted separate volume cert-volume in deployment.yaml
- Updated config
- Update README.md
- Updated version

tlawrie commented:
> @tlawrie What's your thoughts on how to best handle replicaSets and other parameters?

> What do you mean, handle replicaSets and other parameters?

Hey @gchickma just checking on the above.

tlawrie commented on bmrg-flow/values.yaml:
```diff
@@ -108,6 +108,10 @@ global:
       password: passw0rd
       secretName:
+      extraParams:
+      tls:
+        enable: false
+        cert:
```
Two questions

1. Can we just point to a secret rather than uploading the cert? I.e. I am guessing when MongoDB installs, it has the crt published in a secret?

2. Can we combine this to be tlsSecretName and just reference that?

gchickma replied:
@tlawrie Whats your thoughts on the above?

tlawrie replied:
So a few things here.

1. This section (and the database parameters) are in the global section which is pushed / populated to all sub charts. We want to try and keep this as slim line as possible.

2. What is the contents of the cert? Its a TLS certificate yeah? 

3. If the MongoDB is installed in cluster, does it generate the cert and stick it in the same secret as the base64 encoded username and password? Or do you still have to manually upload it?

4. If installed outside cluster, what details do you need? is it just TLS cert or do you still need username and password?

tlawrie commented:
Hi @gchickma following up on the above review. Thanks.

tlawrie commented:
Hi @gchickma following up on the review. Thanks
